### PR TITLE
Support compatibility for IDP role to local role mappings behaviour in IDP group to role assignments

### PR DESCRIPTION
**Pull request by SujanSanjula96**
### Proposed changes in this pull request
This PR contains improvements to IDP group to local role assignment feature to support compatibility for IDP role to local role mappings behaviour. This PR,
- Add the support for sending unmapped IDP groups as roles. (This is to preserve the capability of sending unmapped roles in IDP role to local role mappings)
- Use the IDP role claim URI as the IDP group claim URI. (This is to prevent changing the current IDP claim mappings and issues when using local claim dialect.)
- Handle the above scenarios in JIT provisioning flow
- Add support for SP role mapping for federated users
- Handle the role claim in DefaultClaimHandler

Issue - https://github.com/wso2/product-is/issues/18714

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java [hunk 1]
```diff
@@ -179,12 +190,67 @@ protected Map<String, String> handleFederatedClaims(Map<String, String> remoteCl
...
         remoteClaims.putAll(context.getRuntimeClaims());
 
+        // This handles the roles claim of the federated user.
+        List<String> federatedUserRolesUnmappedInclusive;
+        // This handles the mapped roles of the federated user using for the scope validation.
+        List<String> federatedUserRolesUnmappedExclusive = new ArrayList<>();
+        // This handle the sp mapped roles.
+        String serviceProviderMappedUserRoles;
+
+        boolean useIdpRoleLocalRoleMapping = CarbonConstants.ENABLE_LEGACY_AUTHZ_RUNTIME && !isAppRoleResolverExists();
```
Comment (AnuradhaSK) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java:200]:
```suggestion
        boolean useIdpRoleLocalToRoleMapping = CarbonConstants.ENABLE_LEGACY_AUTHZ_RUNTIME && !isAppRoleResolverExists();
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java [hunk 2]
```diff
@@ -267,16 +326,115 @@ protected Map<String, String> handleFederatedClaims(Map<String, String> remoteCl
...
+        if (!useIdpRoleLocalRoleMapping) {
+            if (CollectionUtils.isNotEmpty(federatedUserRolesUnmappedExclusive)) {
+                // Adding identity provider mapped user roles to be used in the federated user role resolver
+                // for scope validation.
+                spFilteredClaims.put(FrameworkConstants.IDP_MAPPED_USER_ROLES,
+                        String.join(FrameworkUtils.getMultiAttributeSeparator(), federatedUserRolesUnmappedExclusive));
+                spFilteredClaims.put(FrameworkConstants.APP_ROLES_CLAIM,
+                        String.join(FrameworkUtils.getMultiAttributeSeparator(),
+                                federatedUserRolesUnmappedExclusive));
+
```
Comment (AnuradhaSK) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java:343]:
```suggestion
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java [hunk 3]
```diff
@@ -267,16 +326,115 @@ protected Map<String, String> handleFederatedClaims(Map<String, String> remoteCl
...
+            } else {
+                spFilteredClaims.put(FrameworkConstants.IDP_MAPPED_USER_ROLES, StringUtils.EMPTY);
+            }
+
```
Comment (AnuradhaSK) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java:347]:
```suggestion
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java (hunk 1 above)
Comment (SujanSanjula96) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java]: Fixed with https://github.com/wso2/carbon-identity-framework/pull/5396/commits/20cda0fb2dd65caf25f249b337c907124de9280d

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java (hunk 3 above)
Comment (SujanSanjula96) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java]: Fixed with https://github.com/wso2/carbon-identity-framework/pull/5396/commits/20cda0fb2dd65caf25f249b337c907124de9280d

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java [hunk 4]
```diff
@@ -267,16 +327,124 @@ protected Map<String, String> handleFederatedClaims(Map<String, String> remoteCl
...
+     * Map the external IDP roles to local roles.
+     * If excludeUnmapped is true exclude unmapped roles.
+     * Otherwise include unmapped roles as well.
+     *
+     * @param externalIdPConfig     External IDP config.
+     * @param extAttributesValueMap External IDP attributes.
+     * @param idpRoleClaimUri       IDP role claim URI.
+     * @param excludeUnmapped       Whether to exclude unmapped roles.
+     * @return ArrayList<string> List of identity provider mapped user roles.
+     */
```
Comment (AnuradhaSK) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java:372]: exception is not documented in the comment

**Review by AnuradhaSK — DISMISSED** [review]

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java (hunk 4 above)
Comment (SujanSanjula96) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java:373]: Fixed with

**Review by AnuradhaSK — APPROVED** [review]

**Pull request merged**